include/seastar/core/semaphore.hh

```diff
@@ -214,6 +232,30 @@ public:
     {
         static_assert(std::is_nothrow_move_constructible_v<expiry_handler>);
     }
+#ifdef SEASTAR_SEMAPHORE_DEBUG
+    basic_semaphore() noexcept(std::is_nothrow_default_constructible_v<exception_factory>)
+        : basic_semaphore(0)
+    {}
+    basic_semaphore(basic_semaphore&& o) noexcept(std::is_nothrow_move_constructible_v<exception_factory>)
+        : exception_factory(std::move(o))
+        , _count(std::exchange(o._count, 0))
+        , _ex(std::move(o._ex))
+        , _wait_list(std::move(o._wait_list))
+        , _used(o._used)
+    {
+        assert(!_used && "semaphore cannot be moved after it has been used");
+    }
```

Comment:
You can have a class used_flag { bool used } to avoid writing the complex constructors/assignops and focus only on the used flag. It will reduce complexity and duplication.

Reply:
done in v4